rfcs/0004-checkpoints.md

```diff
@@ -0,0 +1,315 @@
+    /// Defaults to 64MB
+    pub max_memtable_bytes: u64
+}
+
+
+struct DbReader {
+    …
+}
+
+impl DbReader {
+    /// Creates a database reader that can read the contents of a database (but cannot write any
+    /// data). The caller can provide an optional checkpoint. If the checkpoint is provided, the
+    /// reader will read using the specified checkpoint and will not periodically refresh the checkpoint. Otherwise, the reader creates a new checkpoint pointing to the current manifest and refreshes it periodically as specified in the options.
+    pub async fn open(
+        path: Path,
+        object_store: Arc<dyn ObjectStore>,
+        checkpoint: Option<UUID>,
+        options: DbReaderOptions
+    ) -> Result<Self, SlateDBError> {...}
+
+    /// Read a key an return the read value, if any.  
+  pub async fn get(&self, key: &[u8]) -> Result<Option<Bytes>, SlateDBError> {...}
+}
+```
+
+Finally, we’ll also allow users to create checkpoints using the admin CLI:
+
+```bash
+$ slatedb create-checkpoint --help
+Create a new checkpoint pointing to the database's current state
+
+Usage: slatedb --path <PATH> create-checkpoint [OPTIONS]
+
+Options:
+  -l, --lifetime <LIFETIME>  Optionally specify a lifetime for the created checkpoint. You can specify the lifetime in a human-friendly format that uses years/days/min/s, e.g. "7days 30min". The checkpoint's expiry time will be set to the current wallclock time plus the specified lifetime. If the lifetime is not specified, then the checkpoint is set with no expiry and must be explicitly removed
+
+$ slatedb refresh-checkpoint --help
+Refresh an existing checkpoint's expiry time. This command will look for an existing checkpoint and update its expiry time using the specified lifetime
+
+Usage: slatedb --path <PATH> refresh-checkpoint [OPTIONS] --id <ID>
+
+Options:
+  -i, --id <ID>              The UUID of the checkpoint (e.g. 01740ee5-6459-44af-9a45-85deb6e468e3)
+  -l, --lifetime <LIFETIME>  Optionally specify a new lifetime for the checkpoint. You can specify the lifetime in a human-friendly format that uses years/days/min/s, e.g. "7days 30min". The checkpoint's expiry time will be set to the current wallclock time plus the specified lifetime. If the lifetime is not specified, then the checkpoint is updated with no expiry and must be explicitly removed
+
+$ slatedb delete-checkpoint --help
+Delete an existing checkpoint
+
+Usage: slatedb --path <PATH> delete-checkpoint --id <ID>
+
+Options:
+  -i, --id <ID>  The UUID of the checkpoint (e.g. 01740ee5-6459-44af-9a45-85deb6e468e3)
+```
+
+### Writers
+
+Writers will create checkpoints to “pin” the set of SRs currently being used to serve reads. As the writer consumes manifest updates, it will re-establish its checkpoint to point to the latest manifest version. Eventually, when we support range scans, the writer will retain older checkpoints to ensure that long-lived scans can function without interference from the GC.
```

Comment:
Would the writer always persist the checkpoint for range scans? I think you were proposing to keep it only in memory.
Also, wouldn't range scans use sequence number like mechanism instead of checkpoints, so that it can i) Avoid a write for every range scan and ii) Read uncommitted data in range scans.

Reply:
> Would the writer always persist the checkpoint for range scans? I think you were proposing to keep it only in memory.

The writer always maintains a checkpoint in the manifest for its current view of the database. I was thinking the range scans could hold references to the checkpoint and then the writer can drop the checkpoint when all references have dropped.

> Also, wouldn't range scans use sequence number like mechanism instead of checkpoints, so that it can i) Avoid a write for every range scan and ii) Read uncommitted data in range scans.

The range scans would still use a sequence number to isolate the scan from concurrent writes. The intent of the checkpoint is to protect the scan from the garbage collector. This doesn't require a write for every range scan. The writer maintains a checkpoint for its current view of the manifest, and range scans will use and share this checkpoint. Uncommitted data is read directly from the in-memory table so it doesn't need to be tied to a checkpoint